### DISCUSSION TITLE
Multiple platforms for OAuth Support?

### DISCUSSION THREAD

**Discussion by chrisdhanaraj** [supabase/supabase#4220] [Questions] [answered]
Hello there! 

Using Google as an example, Google OAuth wants me to create a new client for each platform I'm trying to target; i.e., I have a client for iOS, Android, and Web. In Supabase, I only see the space to add one client, versus Auth0 which allows me to create multiple different applications. 

Does Supabase Auth support multiple platforms, or is it in the works?

**Comment by J0** [comment]
Hey, unfortunately, for the case of Google we only support one client per OAuth provider(e.g. web/iOS/Android) for the moment. It's something that we plan on revisiting in the near future though.

**Reply by dmassiani** [reply]
Any update about this feature?

**Comment by kangmingtay** [comment, marked as answer]
@chrisdhanaraj @dmassiani there is now support for multiple google client IDs - you can visit your supabase project auth settings and under the google provider, there should be an "Authorized Client IDs (for Android, One Tap, and Chrome extensions)" option

![image](https://github.com/supabase/supabase/assets/28647601/badd2917-8789-46bf-9a7d-adac615d5ec3)

**Reply by dmassiani** [reply]
@kangmingtay You are amazing! That's an absolute good news for me! Thank you ❤️

**Reply by elyobo** [reply]
@kangmingtay how can we set this in the `config.toml` for local development?

**Reply by fasidOnGit** [reply]
But the secret remains the same ? For 2 different Client Ids i got 2 different client secrets. But supabase accepts multiple client ids but only one secret ?

**Reply by elyobo** [reply]
If you're doing sign in on the web then you (almost certainly) don't need multiple client IDs and secrets. If you do, for some reason, then you'll need to implement the sign in flow yourself (and have the signing in chose which client ID they want to sign in with I guess) and put the client IDs in the authorised client IDs section.

If you're not doing sign in on the web, e.g. you're only doing mobile sign in, you only need to use the authorised client IDs section, which accepts multiple IDs (but you still probably only need one ID there most of the time, the web ID, neither the Android nor the iOS key needs to be in there).

**Comment by wishup-agrit** [comment]
Hey @kangmingtay , still facing the error while I added the client ID for iOS in new field. Has anyone tried using the google provider for two client IDs  one of which is iOS.

**Comment by baronha** [comment]
Does anyone have any updates for Supabase CLI that can use multiple client_id at the same time for `config.toml` like this:
``` config.tomp
client_ids = "env(GOOGLE_IOS_CLIENT_ID),env(GOOGLE_WEB_CLIENT_ID)"
```

**Reply by elyobo** [reply]
You don't need to, you only need the web client ID in there even to do iOS google sign in.

**Reply by elyobo** [reply]
But also I have seen it commented elsewhere in discussions on this issue that comma separated values do work, if you do actually need them for some reason (like having multiple web IDs).

**Reply by baronha** [reply]
OH God. This actually works on iOS. I spent all day testing yesterday but it didn't work. Thank you very much

**Comment by masfworld** [comment]
I'm using React Native + Expo to develop an application for web and iOS (and potentially Android) I have the Google OAuth working in Web, but not in iOS, is it compatible with Supabase? Any example?

Thanks in advance

**Reply by elyobo** [reply]
We have an react native + expo app with Google OAuth which works across iOS and Android. It's necessary to create three OAuth client IDs, one for web, one for iOS, and one for Android. Only the web client ID needs to be set in the Supabase config for Google auth, the web and the iOS client IDs are needed in your build, and the Android one just needs to exist in your Google account. This is not documented well, as far as I can tell - it was very confusing when I set it up.

**Reply by masfworld** [reply]
Yep, it's not very well documented. I decided using `GoogleSignIn` library for Android and iOS and supabase API for Web. Using that library means that I cannot use Expo Go anymore, I'm using `EAS` now. I've decided to postpone iOS implementation because it requires the developer license right now.

This is my code. It's working for Web and Android. Happy to receive feedback
```js
import { createContext, useContext, useEffect, useState } from 'react'
import { Session } from '@supabase/supabase-js'
import { supabase } from '../utils/supabase'
import { Platform } from 'react-native'
import { router } from 'expo-router'
import { InteractionManager } from 'react-native'
import {
  GoogleSignin,
  statusCodes,
  isErrorWithCode,
  isSuccessResponse,
} from '@react-native-google-signin/google-signin'

// Define the shape of our auth context
type AuthContextType = {
  session: Session | null        // Current session state
  loading: boolean              // Loading state for auth operations
  signInWithGoogle: () => Promise<void>  // Google sign-in function
  signOut: () => Promise<void>  // Sign-out function
}

// Create the auth context with undefined as initial value
const AuthContext = createContext<AuthContextType | undefined>(undefined)

export function AuthProvider({ children }: { children: React.ReactNode }) {
  // State for managing session and loading status
  const [session, setSession] = useState<Session | null>(null)
  const [loading, setLoading] = useState(true)
  

  useEffect(() => {
    // Flag to prevent state updates after component unmount
    let isMounted = true;

    const initSession = async () => {
       // Configure Google Sign-In for Android
      if (Platform.OS === 'android') {
        await GoogleSignin.configure({
          webClientId: process.env.EXPO_PUBLIC_GOOGLE_WEB_CLIENT_ID, // Web Client ID from Google Cloud Console
        });
      }
      try {
        // Get the initial session from Supabase
        const { data: { session }, error } = await supabase.auth.getSession()
        if (error) throw error
        
        // Update session state only if component is still mounted
        if (isMounted) {
          await InteractionManager.runAfterInteractions(() => {
            setSession(session)
            setLoading(false)
          })
        }
      } catch (error) {
        console.error('Error getting session:', error)
      }
    }

    initSession()

    // Subscribe to auth state changes
    const { data: { subscription } } = supabase.auth.onAuthStateChange((_event, session) => {
      if (isMounted) {
        InteractionManager.runAfterInteractions(() => {
          setSession(session)
        })
      }
    })

    // Cleanup function
    return () => {
      isMounted = false
      subscription.unsubscribe()
    }
  }, [])

  // Handle Google Sign In
  const signInWithGoogle = async () => {
    try {
      // Web platform authentication
      if (Platform.OS === 'web') {
        const { error } = await supabase.auth.signInWithOAuth({
          provider: 'google',
          options: {
            redirectTo: process.env.URL,
          },
        })
        if (error) throw error
      } else {
        // Native platforms authentication
        try {
          // Check if Google Play Services are available (Android)
          await GoogleSignin.hasPlayServices();
          // Trigger native Google Sign In flow
          const response = await GoogleSignin.signIn();
          console.log(JSON.stringify(response, null, 2))
          if (isSuccessResponse(response)) {
            // Sign in to Supabase with the Google ID token
            const { data, error } = await supabase.auth.signInWithIdToken({
              provider: 'google',
              token: response.data.idToken!
            });
            if (error) throw error
            router.replace('/')
          } else {
            console.log('Auth cancelled by the user')
          }
        } catch (error) {
          // Handle specific Google Sign In errors
          if (isErrorWithCode(error)) {
            switch (error.code) {
              case statusCodes.IN_PROGRESS:
                console.log('Operation already in progress')
                break;
              case statusCodes.PLAY_SERVICES_NOT_AVAILABLE:
                console.log('Play services not available or outdated')
                break;
              default:
                console.log('Google Signin Error')
            }
          } else {
            console.log('Google Auth Error')
          }
        }
      }
    } catch (error) {
      console.error('Error in signInWithGoogle:', error)
    }
  }

  // Handle Sign Out
  const signOut = async () => {
    try {
      // Sign out from Supabase
      const { error } = await supabase.auth.signOut()
      if (error) throw error

      // Sign out from Google (Android only)
      if (Platform.OS === 'android') {
        await GoogleSignin.signOut()
      }
      
      // Clear session state after all interactions are complete
      await InteractionManager.runAfterInteractions(() => {
        setSession(null)
      })
    } catch (error) {
      console.error('Error in signOut:', error)
    }
  }

  // Provide auth context to children components
  return (
    <AuthContext.Provider value={{ session, loading, signInWithGoogle, signOut }}>
      {children}
    </AuthContext.Provider>
  )
}

// Custom hook to use auth context
export const useAuth = () => {
  const context = useContext(AuthContext)
  if (context === undefined) {
    throw new Error('useAuth must be used within an AuthProvider')
  }
  return context
}


```

**Reply by TasseDeCafe** [reply]
> We have an react native + expo app with Google OAuth which works across iOS and Android. It's necessary to create three OAuth client IDs, one for web, one for iOS, and one for Android. Only the web client ID needs to be set in the Supabase config for Google auth, the web and the iOS client IDs are needed in your build, and the Android one just needs to exist in your Google account. This is not documented well, as far as I can tell - it was very confusing when I set it up.

Damn, the fact that I had to find some obscure GH issues thread from 4 years ago to have this spelled out for me is crazy. Thanks.

**Reply by nikolaiessel** [reply]
> We have an react native + expo app with Google OAuth which works across iOS and Android.

@elyobo But I think not with a simple supabase call? On Web and Android its clear and the answer of @masfworld contains everything needed.
But on iOS there is no native stuff of google and signInWithOAuth is designed for web only, requiring a browser (which is not clear documented):
```
    if (isBrowser() && !options.skipBrowserRedirect) {
      window.location.assign(url)
    }

    return { data: { provider, url }, error: null }
```

**Reply by elyobo** [reply]
@nikolaiessel, @masfworld's answer uses `@react-native-google-signin/google-signin` which provides the necessary functionality for `iOS` as well. Incomplete code snippets below, the first providing the `googleSignInHandler` referenced by the second which is one of the sign in UIs. It works across iOS and Android, we don't build for web. The gist is that you get an ID token from the Google sign in flow which you then pass to Supabase to validate, which it can using the configured web client ID provided to it per [my answer above](https://github.com/orgs/supabase/discussions/4220#discussioncomment-11780758).

```ts
import { Alert } from 'react-native'
import {
  GoogleSignin as GoogleSignInApi,
  statusCodes,
  type User as GoogleUser,
} from '@react-native-google-signin/google-signin'

// NOTE(@elyobo): We only want users with ID tokens
type GoogleUserWithToken = Omit<GoogleUser, 'idToken'> & { idToken: string }

export const googleSignInHandler =
  (handleRequest: (userInfo: GoogleUserWithToken) => void | Promise<void>) =>
  async () => {
    GoogleSignInApi.configure({
      iosClientId: process.env.EXPO_PUBLIC_GOOGLE_IOS_CLIENT_ID,
      webClientId: process.env.EXPO_PUBLIC_GOOGLE_WEB_CLIENT_ID,

      forceCodeForRefreshToken: true,
      offlineAccess: false,
      profileImageSize: 400,
      scopes: ['profile', 'email'],
    })

    try {
      await GoogleSignInApi.hasPlayServices()
      if (GoogleSignInApi.hasPreviousSignIn()) await GoogleSignInApi.signOut()
      const userInfo = await GoogleSignInApi.signIn()

      if (!userInfo.idToken) throw new Error('no ID token present!')
      // NOTE(@elyobo): ID token presence is validated above
      await handleRequest(userInfo as GoogleUserWithToken)

      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } catch (error: any) {
      // handle non-cancellation errors, cancellation is fine
      if (error.code !== statusCodes.SIGN_IN_CANCELLED)
        Alert.alert('Error', error.message)
    }
  }
```

```ts
import { Alert } from 'react-native'

import GoogleIcon from '@/assets/svgs/google-icon'
import { useSupabase } from '@/components/SessionProvider'
import { router } from '@/helpers/typedRouter'

import { googleSignInHandler } from './googleSignIn'
import SignInButton from './SignInButton'

const GoogleSignInButton = () => {
  const supabase = useSupabase()
  return (
    <SignInButton
      icon={GoogleIcon}
      label="Continue with Google"
      onPress={googleSignInHandler(async userInfo => {
        const { error } = await supabase.auth.signInWithIdToken({
          provider: 'google',
          token: userInfo.idToken,
        })
        if (error) {
          Alert.alert('Error', error.message)
          return
        }
        router.replace('/')
      })}
    />
  )
}

export default GoogleSignInButton
```

**Reply by nikolaiessel** [reply]
@elyobo Thanks for follow up. The issue was I missinterpreted the thrown error as "Google services not available on iOS". But instead your inital comment contained the shortend solution:
- Create an additional OAuth client in google developer console for iOS
- Add the client id to the supabase google auth provider and disable Nonce-Check
- Add the client to the [app.config](https://react-native-google-signin.github.io/docs/setting-up/expo#add-config-plugin)
```
{
    ....
    "plugins": [
        ....
        ["@react-native-google-signin/google-signin", { iosUrlScheme: googleIosClientId.split(".").reverse().join(".") ]
    ]
}
```
- Use the iOS client id for GoogleSignin for iOS
```
	GoogleSignin.configure({
		scopes: [
			"https://www.googleapis.com/auth/userinfo.email",
			"https://www.googleapis.com/auth/userinfo.profile",
			"openid"
		],
		iosClientId: config.auth.googleIosClientId
	});
```

**Comment by harshalk215** [comment]
I am building 2 web platforms. One for coaches and one for clients (with different urls). But basically I need single backend, or shared DB. So I must use single Single Supabase project. Both platforms use Google OAuth and works fine with single GCP project and single client. But I cannot distinguish if signed in user is client or coach as in Supabase both are inserted into auth.users. I thought creating separate clients in GCP OAuth config will give some solution, but I am not able to configure 2 web clients in single Supabase project. So if this comment https://github.com/orgs/supabase/discussions/4220#discussioncomment-12512094 by @elyobo is valid then I am stuck. I am putting my use case here, to just inform that there might be use case of having 2 web platform with different user base but still shared db (ie single supabase project). @elyobo also mentioned this https://github.com/orgs/supabase/discussions/4220#discussioncomment-11768821 , but I am not clear if it works or not. How does just supporting multiple client ids but not accepting all secrets even be useful.

**Reply by elyobo** [reply]
From my comment that you referenced

> If you do, for some reason [need multiple client id / secrets], then you'll need to implement the sign in flow yourself

I think you need to sign them in yourself, rather than e.g. using the supabase redirect URL, like [how we do on mobile](https://github.com/orgs/supabase/discussions/4220#discussioncomment-13577546) and then call `supabase.auth.signInWithIdToken()` as in that example. In that case Supabase doesn't need the secret at all, it's just verifying the token that Google has provided, and that it was generated using one of the client IDs you've provided.

That said, then you've still just signed in the user, you can't tell which OAuth client ID they used to sign in with - and unless there are somehow some restrictions that you're applying to limit who can sign in with which client IDs, I'm not clear how it's providing you any security about your two types of users either.

It seems like you could use the single client/secret and track which type of user they are in your application instead? We actually have two types of users in our app as well, but have two different sign in flows, one of which just signs in with Google and the other which signs them in with Google and then also verifies their access in another API of ours via a Supabase edge function which writes some extra metadata the DB to indicate their user type. Then we use a [custom access token hook](https://supabase.com/docs/guides/auth/auth-hooks/custom-access-token-hook) to verify that their access is still valid, and to modify the Supabase auth token to tell us which type of user they are so that the client side functionality can behave appropriately.

**Reply by harshalk215** [reply]
Thanks @elyobo for detailed explaination.
Yes, now I am handling it in my app itself. Basically I just insert user in coach or client table and that's why I needed distinction. Doing it in Supabase immediately with insert trigger would have been robust. But as you suggested, I am handling it in my app now, both app codebases are anyhow different, so after successful signup/login, I will upsert in appropriate tables. Using Supabase for the first time, but glad to get a prompt response.
